packages/vite/src/node/plugins/resolve.ts

```diff
@@ -658,14 +658,25 @@ function mapWithBrowserField(
   relativePathInPkgDir: string,
   map: Record<string, string | false>
 ): string | false | undefined {
-  const normalized = normalize(relativePathInPkgDir)
+  const normalizedPath = path.posix.normalize(relativePathInPkgDir)
+
   for (const key in map) {
-    if (normalize(key) === normalized) {
+    const normalizedKey = path.posix.normalize(key)
+    const isJsExt = normalizedKey.endsWith('.js')
+    const isIndexJs = normalizedKey.endsWith('/index.js')
+
+    const substitutions: string[] = [normalizedKey]
+
+    if (isJsExt) {
+      substitutions.push(normalizedKey.slice(0, -'.js'.length))
+    }
+
+    if (isIndexJs) {
+      substitutions.push(normalizedKey.slice(0, -'/index.js'.length))
+    }
+
+    if (substitutions.includes(normalizedPath)) {
       return map[key]
     }
   }
 }
-
```

Comment:
What do you think about extracting a helper to simplify this logic so there is no need for the creation of an array?
```ts
function mapWithBrowserField(
  relativePathInPkgDir: string,
  map: Record<string, string | false>
): string | false | undefined {
  const normalizedPath = path.posix.normalize(relativePathInPkgDir)
  for (const key in map) {
    const normalizedKey = path.posix.normalize(key)
    if (
      normalizedPath === normalizedKey ||
      equalWithoutSuffix(normalizedPath, normalizedKey, '.js') ||
      equalWithoutSuffix(normalizedPath, normalizedKey, '/index.js')
    ) {
      return map[key]
    }
  }
}

function equalWithoutSuffix(path: string, key: string, suffix: string) {
  return path === key.endsWith(suffix) && key.slice(0, -suffix.length)
}
```

Reply:
👍🏻, done.